File: tree/TreeLongestPath.cpp

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        maxDm = 0;
        getMaxDepth(root);
        return maxDm;
    }

private:
    int maxDm;
    int getMaxDepth(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        int leftDm = getMaxDepth(root->left);
        int rightDm = getMaxDepth(root->right);
        maxDm = max(leftDm + rightDm, maxDm);
        return 1 + max(leftDm, rightDm);
    }
};

File: tree/TreeLongestPath.java

public class TreeLongestPath {
    private int maxDm = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        getMaxDepth(root);
        return maxDm;
    }

    private int getMaxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftLen = getMaxDepth(root.left);
        int rightLen = getMaxDepth(root.right);
        maxDm = Math.max(maxDm, leftLen + rightLen);
        return 1 + Math.max(leftLen, rightLen);
    }
}
